src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -166,6 +183,47 @@ public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> ge
             }
         }
 
+        public IEnumerable<KeyValuePair<string, string>> ExtractHeaderTags<T>(T headers, IEnumerable<KeyValuePair<string, string>> headerToTagMap, string defaultTagPrefix)
+            where T : IHeadersCollection
+        {
+            foreach (KeyValuePair<string, string> headerNameToTagName in headerToTagMap)
+            {
+                string headerValue = ParseString(headers, headerNameToTagName.Key);
+                if (headerValue is null)
+                {
+                    continue;
+                }
+
+                // Tag name is normalized during Tracer instantiation so use as-is
+                if (!string.IsNullOrWhiteSpace(headerNameToTagName.Value))
+                {
+                    yield return new KeyValuePair<string, string>(headerNameToTagName.Value, headerValue);
+                }
+                else
+                {
+                    // Since the header name was saved to do the lookup in the input headers,
+                    // convert the header to its final tag name once per prefix
+                    var cacheKey = new Key(headerNameToTagName.Key, defaultTagPrefix);
+                    string tagNameResult = DefaultTagMappingCache.GetOrAdd(cacheKey, key =>
+                    {
+                        if (key.HeaderName.TryConvertToNormalizedHeaderTagName(out string normalizedHeaderTagName))
+                        {
+                            return key.TagPrefix + "." + normalizedHeaderTagName;
+                        }
+                        else
+                        {
+                            return null;
+                        }
+                    });
+
+                    if (!string.IsNullOrWhiteSpace(tagNameResult))
```

Comment:
nit: I guess `tagNameResult` is always either `null` or non-empty, so could probably replace with a null check here?

Reply:
Yeah good catch, I'll update that

Reply:
Done in b3f9c8c9975f0bbf1ada23a363a69edf4e77ac8d